src/ordered_dict.jl

```diff
@@ -24,19 +24,34 @@ type OrderedDict{K,V} <: Associative{K,V}
     function OrderedDict()
         new(zeros(Int32,16), Array(K,0), Array(V,0), 0, false)
     end
+    function OrderedDict(n::Integer)
+        slotsz = max(16, (n*3)>>1 )
+        new(zeros(Int32,slotsz), Array(K,n), Array(V,n), 0, false)
+    end
     function OrderedDict(kv)
-        h = OrderedDict{K,V}()
+        h = OrderedDict{K,V}(length(kv))
```

Comment:
There already is `Base.HasLength()`: https://github.com/JuliaLang/julia/blob/bcc2121dc31647fc0999d09c10df91d35f216838/base/generator.jl#L44